### DISCUSSION TITLE
Does uBO de-dupe cosmetic-HTML filters if no-script?

### DISCUSSION THREAD

**Discussion by Rudxain** [uBlockOrigin/uBlock-issues#3395] [General]
If `no-scripting: example.com true`, and we have an HTML-filter such as:
```adblock
example.com##^.ad-unit-wrapper
```
Then the following cosmetic-filters are useless for `example.com`:
```adblock
! generic: has to stay in memory, but
! `example.com` can be whitelisted
##.ad-unit-wrapper
! specific: can be deduped
example.com##.ad-unit-wrapper
! very specific: not identical, but can be discarded anyways
example.com##.ad-unit-wrapper > .example-ad
```
This optimization is valid for all host-names, as long as the HTML-filter's CSS-selector is identical (or "broader") than the cosmetic counterpart.

I understand that static-filters are parsed+de-duped independently of rules (dynamic or switch). But what if the user [blocks all scripts using a static-filter](https://github.com/uBlockOrigin/uBlock-issues/discussions/3395#discussioncomment-10820090)?

**Comment by garry-ut99** [comment]
> But what if the user blocks all scripts using a static-filter?

`*$script,inline-script,domain=example.com` gives the same effect.

**Reply by krystian3w** [reply]
#3181 but can be fixed by recommend CSP.

**Comment by gorhill** [comment]
I don't understand what you are asking. Cosmetic filters are unrelated to `no-scripting` switch.

**Reply by krystian3w** [reply]
It seems that he would suggest checking if JS is inactive and if we have HTML filtering match, then we should not load CSS filters because the page will not dynamically replace the content due to lack of access to JS.

**Reply by Rudxain** [reply]
> Cosmetic filters are unrelated to `no-scripting`

That's what I meant by "independent". So I mentioned static-filters as an alternative, because I assume it's easier to implement the optimization when the script-blocking is specified by a [filter](https://github.com/uBlockOrigin/uBlock-issues/discussions/3395#discussioncomment-10820090) rather than a rule (switch and/or dynamic).

> if JS is inactive and if we have HTML filtering match, then we should not load CSS filters

**Spot-on!**

**Reply by stephenhawk8054** [reply]
> Does uBO de-dupe cosmetic-HTML filters if no-script?

Then no.

**Reply by gorhill** [reply]
> if JS is inactive and if we have HTML filtering match, then we should not load CSS filters
> 
> Spot-on!

Please provide profiling data showing that the "gain" -- if measurable at all -- is worth the amount of work to invest in making this happen, the added pile of JS code paths, and the added overhead of having to verify that all the conditions are fulfilled to avoid injecting the CSS filters.

**Comment by krystian3w** [comment]
CSS can be useful if HTML Filtering fails on the first visit https://github.com/uBlockOrigin/uBlock-issues/issues/3222 (I assume that disabling JS by any method does not help to reported error 3222).
